README.md
# React SVG Brush (Beta)

The [react-svg-brush](https://www.npmjs.com/package/react-svg-brush) is a React-based brush library that emulates the [d3-brush](https://github.com/d3/d3-brush) behavior. It renders the brush using React virtual DOM.

This library is still in Beta, please file bug reports by creating issues in the library [github repo](https://github.com/kenns29/react-svg-brush).

## Examples

**1D Brush**

[<img alt="brush-1d" src="docs/brush-1d.png"></img>](https://codepen.io/kennsmacintosh/pen/yLBPGbN)

**2D Brush**

[<img alt="brush-1d" src="docs/brush-2d.png" width="200px"></img>](https://codepen.io/kennsmacintosh/pen/dybZwQj)

**Brush Snapping (Transition)**

[<img alt="brush-1d" src="docs/brush-1d.png"></img>](https://codepen.io/kennsmacintosh/pen/oNvoVaY)

## Why Yet Another Brush Library

The [d3-brush](https://github.com/d3/d3-brush) library is commonly used for data selection within visualizations. It enables data selection through dragging on the visualization interface (holding the left mouse and move the curser). It also renders the brushed area directly on the visualization. However, there are several drawbacks when using [d3-brush](https://github.com/d3/d3-brush) directly inside a React-based application:

- d3-brush directly modifies the DOM, which also means customization of the brush behavior/view has to be done directly on the DOM, violating the best practices in React.
- Using d3-brush inside a React component typically requires using the react lifecycle methods, e.g. `componentDidUpdate`, adding complexity in the implementation of the React component.

The React SVG Brush library emulates the d3-brush library DOM structure (95% identical) and functionalities but renders the Brush completely using the React virtual DOM, making it easy for applications that use d3-brush to switch to this library to adhere to the React best practices.

## Installation

```
npm install react-svg-brush
```

## Example Usage

```javascript
import React, {Component} from 'react';
import SVGBrush from 'react-svg-brush';

...

export default class App extends Component {
  onBrushStart = ({target, type, selection, sourceEvent})=>{...}
  onBrush = ({target, type, selection, sourceEvent})=>{...}
  onBrushEnd = ({target, type, selection, sourceEvent})=>{...}
  renderBrush = () => (
    <SVGBrush
      // Defines the boundary of the brush.
      // Strictly uses the format [[x0, y0], [x1, y1]] for both 1d and 2d brush.
      // Note: d3 allows the format [x, y] for 1d brush.
      extent={[[x0, y0], [x1, y1]]}
      // Obtain mouse positions relative to the current svg during mouse events.
      // By default, getEventMouse returns [event.clientX, event.clientY]
      getEventMouse={event => {
        const {clientX, clientY} = event;
        const {left, top} = this.svg.getBoundingClientRect();
        return [clientX - left, clientY - top];
      }}
      brushType="2d" // "x"
      onBrushStart={this.onBrushStart}
      onBrush={this.onBrush}
      onBrushEnd={this.onBrushEnd}
    />
  )
  render() {
    return (
      <svg ref={input => (this.svg = input)}>
        {this.renderBrush()}
      </svg>
    );
  }
}
```

The following shows the DOM structure generated by React SVG Brush:

```html
<g class="brush">
  <rect class="overlay" pointer-events="all" cursor="crosshair" fill="none" x="40" y="10" width="380" height="350"></rect>
  <rect class="selection" cursor="move" fill="#777" fill-opacity="0.3" stroke="#fff" shape-rendering="crispEdges" x="217.80859375" y="144.046875" width="30.26171875" height="185.546875"></rect>
  <rect class="handle handle--n" cursor="ns-resize" x="212.80859375" y="139.046875" width="40.26171875" height="10" fill="none" pointer-events="visible"></rect>
  <rect class="handle handle--e" cursor="ew-resize" x="243.0703125" y="139.046875" width="10" height="195.546875" fill="none" pointer-events="visible"></rect>
  <rect class="handle handle--s" cursor="ns-resize" x="212.80859375" y="324.59375" width="40.26171875" height="10" fill="none" pointer-events="visible"></rect>
  <rect class="handle handle--w" cursor="ew-resize" x="212.80859375" y="139.046875" width="10" height="195.546875" fill="none" pointer-events="visible"></rect>
  <rect class="handle handle--nw" cursor="nwse-resize" x="212.80859375" y="139.046875" width="10" height="10" fill="none" pointer-events="visible"></rect>
  <rect class="handle handle--ne" cursor="nesw-resize" x="243.0703125" y="139.046875" width="10" height="10" fill="none" pointer-events="visible"></rect>
  <rect class="handle handle--se" cursor="nwse-resize" x="243.0703125" y="324.59375" width="10" height="10" fill="none" pointer-events="visible"></rect>
  <rect class="handle handle--sw" cursor="nesw-resize" x="212.80859375" y="324.59375" width="10" height="10" fill="none" pointer-events="visible"></rect>
</g>
```

More demo code can be found in this repo, [React SVG Brush Example](https://github.com/kenns29/react-svg-brush-example).

## API Reference

### Attributes

`brushType` **(String, Required)**

- Default: `'2d'`

Choose the brush type. Three brush types are supported: `2d`, `x`, and `y` (`y` is still experimental). The `2d` allows dragging on all sides and corners, whereas the `x` and `y` brush only allows dragging on horizontal and vertical directions respectively.

`extent` **(Array, Required)**

- Default: `[[0, 0], [1, 1]]`

Defines the brush boundary. The array `[[x0, y0], [x1, y1]]` defines the bounding box of the brush boundary, where `[x0, y0]` represents the point in top-left corner and `[x1, y1]` represents the point in the bottom-right corner.

`selection` **(Array, Optional)**

- Default: `undefined`

Manually set the selection of the brush. This is typically used when setting the initial state of the brush, or when post processing the brush selections, e.g. brush snapping. Like the `extent`, the `selection` is also defined by a bounding box array `[[x0, y0], [x1, y1]]`. When the `selection` parameter is defined, the brush will be rendered according to the selection. When the `selection` parameter is set to `null`, the brush will be cleared. When the `selection` parameter is `undefined`, the brush  rendering will be automatically controlled during the interactions.  

### Methods

`getEventMouse` **(Function, Optional)**

- Default: `event => [event.clientX, event.clientY]`

Obtain the mouse position given the pointer event. It must return an array of `[x, y]` that specifies the position of the mouse. By default, it returns `[event.clientX, event.clientY]` which is the position relative to the web page. The default behavior of this function is typically overwritten, because it is more common to use the position relative to the container of the brush (e.g. the SVG).

`onBrushStart` **(Function, Optional)**

The callback function when the brushing interaction starts (i.e. mouse click). This function receives one object as the argument, which contains the following attributes:

- `target` - Refers to the brush object itself (the React component).
- `type` - The type of the callback, in this case, it is `start`
- `selection` - The brush selection when the brush starts.
- `sourceEvent` - The React `SyntheticEvent` of the brush, in this case, it is a pointer event.

`onBrush` **(Function, Optional)**

The callback function during brush interaction (i.e. while dragging the brush). This function receives one object as the argument, which contains the following attributes:

- `target` - Refers to the brush object itself (the React component).
- `type` - The type of the callback, in this case, it is `brush`
- `selection` - The changed brush selection during the interaction.
- `sourceEvent` - The React `SyntheticEvent` of the brush, in this case, it is a pointer event.

`onBrushEnd` **(Function, Optional)**

The callback function at the end of brush interaction (i.e. mouse up). This function receives one object as the argument, which contains the following attributes:

- `target` - Refers to the brush object itself (the React component).
- `type` - The type of the callback, in this case, it is `end`
- `selection` - The changed brush selection at the end of the brush interaction.
- `sourceEvent` - The React `SyntheticEvent` of the brush, in this case, it is a pointer event.

## Change Log

Refer to the [CHANGELOG.md](CHANGELOG.md) file.

CHANGELOG.md
# v0.0.10

- Fix the bug in which the brush is not cleared when the `selection` parameter is set as `null`

# v0.0.9

- Moved react to peer dependencies


src/index.js
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

class SVGBrush extends PureComponent {
  static defaultProps = {
    extent: [
      [0, 0],
      [1, 1]
    ],
    onBrushStart: event => {},
    onBrush: event => {},
    onBrushEnd: event => {},
    getEventMouse: event => [event.clientX, event.clientY],
    brushType: '2d' // 'x', 'y'
  };

  static propTypes = {
    selection: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)),
    extent: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,
    onBrushStart: PropTypes.func.isRequired,
    onBrush: PropTypes.func.isRequired,
    onBrushEnd: PropTypes.func.isRequired,
    getEventMouse: PropTypes.func.isRequired,
    brushType: PropTypes.string.isRequired
  };

  static getDerivedStateFromProps = (props, state) => ({
    ...state,
    selection: props.selection === undefined ? state.selection : props.selection
  });

  constructor(props) {
    super(props);
    this.state = {
      selection: null
    };
    this.move = null;
  }

  _handleBrushStart = event => {
    event.target.setPointerCapture(event.pointerId);
    this.move = this.props.getEventMouse(event);
    this.props.onBrushStart({
      target: this,
      type: 'start',
      selection: this.state.selection,
      sourceEvent: event
    });
  };

  _handleBrushEnd = event => {
    this.move = null;
    this.props.onBrushEnd({
      target: this,
      type: 'end',
      selection: this.state.selection,
      sourceEvent: event
    });
  };

  _renderOverlay() {
    const {
      extent: [[x0, y0], [x1, y1]],
      brushType
    } = this.props;

    return (
      <rect
        className="overlay"
        pointerEvents="all"
        cursor="crosshair"
        fill="none"
        x={x0}
        y={y0}
        width={x1 - x0}
        height={y1 - y0}
        onPointerDown={this._handleBrushStart}
        onPointerMove={event => {
          if (this.move) {
            const [x, y] = this.props.getEventMouse(event);
            const [sx, sy] = this.move;
            let selection = this.state.selection;
            switch (brushType) {
              case '2d':
                selection = [
                  [
                    Math.max(Math.min(sx, x), x0),
                    Math.max(Math.min(sy, y), y0)
                  ],
                  [Math.min(Math.max(sx, x), x1), Math.min(Math.max(sy, y), y1)]
                ];
                break;
              case 'x':
                selection = [
                  [Math.min(sx, x), y0],
                  [Math.max(sx, x), y1]
                ];
                break;
              case 'y':
                selection = [
                  [x0, Math.min(sy, y)],
                  [x1, Math.max(sy, y)]
                ];
            }
            this.setState({selection});
            this.props.onBrush({
              target: this,
              type: 'brush',
              selection,
              sourceEvent: event
            });
          }
        }}
        onPointerUp={event => {
          const move = this.props.getEventMouse(event);
          let selection = this.state.selection;
          if (
            this.move &&
            this.move[0] === move[0] &&
            this.move[1] === move[1]
          ) {
            selection = null;
            this.props.onBrush({
              target: this,
              type: 'brush',
              selection,
              sourceEvent: event
            });
          }
          this.move = null;
          this.setState({selection});
          this.props.onBrushEnd({
            target: this,
            type: 'end',
            selection,
            sourceEvent: event
          });
        }}
      />
    );
  }

  _renderSelection() {
    const {
      extent: [[ex0, ey0], [ex1, ey1]],
      brushType
    } = this.props;
    const {selection} = this.state;
    if (!selection) {
      return null;
    }
    const [[x0, y0], [x1, y1]] = selection;
    const [x, y, w, h] = [x0, y0, x1 - x0, y1 - y0];
    const xbf = x => Math.min(Math.max(x, ex0), ex1);
    const ybf = y => Math.min(Math.max(y, ey0), ey1);
    const sxbf = (x0, x1, dx) => {
      if (x0 + dx < ex0) {
        return [ex0, x1 + (ex0 - x0)];
      }
      if (x1 + dx > ex1) {
        return [x0 + (ex1 - x1), ex1];
      }
      return [x0 + dx, x1 + dx];
    };
    const sybf = (y0, y1, dy) => {
      if (y0 + dy < ey0) {
        return [ey0, y1 + (ey0 - y0)];
      }
      if (y1 + dy > ey1) {
        return [y0 + (ey1 - y1), ey1];
      }
      return [y0 + dy, y1 + dy];
    };

    return (
      <React.Fragment>
        <rect
          className="selection"
          cursor="move"
          fill="#777"
          fillOpacity="0.3"
          stroke="#fff"
          shapeRendering="crispEdges"
          x={x}
          y={y}
          width={w}
          height={h}
          onPointerDown={this._handleBrushStart}
          onPointerMove={event => {
            if (this.move) {
              const [x, y] = this.props.getEventMouse(event);
              const [sx, sy] = this.move;
              const [dx, dy] = [x - sx, y - sy];
              const [mx0, mx1] = sxbf(x0, x1, dx);
              const [my0, my1] = sybf(y0, y1, dy);
              let selection = this.state.selection;
              switch (brushType) {
                case '2d':
                  selection = [
                    [mx0, my0],
                    [mx1, my1]
                  ];
                  break;
                case 'x':
                  selection = [
                    [mx0, y0],
                    [mx1, y1]
                  ];
                  break;
                case 'y':
                  selection = [
                    [x0, my0],
                    [x1, my1]
                  ];
              }
              this.move = [x, y];
              this.setState({selection});
              this.props.onBrush({
                target: this,
                type: 'brush',
                selection,
                sourceEvent: event
              });
            }
          }}
          onPointerUp={this._handleBrushEnd}
        />
        <rect
          ref={input => (this.handleN = input)}
          className="handle handle--n"
          cursor="ns-resize"
          x={x - 5}
          y={y - 5}
          width={w + 10}
          height={10}
          fill="none"
          pointerEvents="visible"
          onPointerDown={this._handleBrushStart}
          onPointerMove={event => {
            if (this.move) {
              const [x, y] = this.props.getEventMouse(event);
              const [sx, sy] = this.move;
              const dy = y - sy;
              const my = ybf(y0 + dy);
              const [my0, my1] = my < y1 ? [my, y1] : [y1, y1];
              let selection = this.state.selection;
              switch (brushType) {
                case '2d':
                case 'y':
                  selection = [
                    [x0, my0],
                    [x1, my1]
                  ];
              }
              this.move = [x, y];
              this.setState({selection});
              this.props.onBrush({
                target: this,
                type: 'brush',
                selection,
                sourceEvent: event
              });
              if (my >= y1) {
                this.handleS.setPointerCapture(event.pointerId);
                return;
              }
            }
          }}
          onPointerUp={this._handleBrushEnd}
        />
        <rect
          ref={input => (this.handleE = input)}
          className="handle handle--e"
          cursor="ew-resize"
          x={x + w - 5}
          y={y - 5}
          width={10}
          height={h + 10}
          fill="none"
          pointerEvents="visible"
          onPointerDown={this._handleBrushStart}
          onPointerMove={event => {
            if (this.move) {
              const [x, y] = this.props.getEventMouse(event);
              const [sx, sy] = this.move;
              const dx = x - sx;
              const mx = xbf(x1 + dx);
              const [mx0, mx1] = x0 < mx ? [x0, mx] : [x0, x0];
              let selection = this.state.selection;
              switch (brushType) {
                case '2d':
                case 'x':
                  selection = [
                    [mx0, y0],
                    [mx1, y1]
                  ];
              }
              this.move = [x, y];
              this.setState({selection});
              this.props.onBrush({
                target: this,
                type: 'brush',
                selection,
                sourceEvent: event
              });
              if (x0 >= mx) {
                this.handleW.setPointerCapture(event.pointerId);
                return;
              }
            }
          }}
          onPointerUp={this._handleBrushEnd}
        />
        <rect
          ref={input => (this.handleS = input)}
          className="handle handle--s"
          cursor="ns-resize"
          x={x - 5}
          y={y + h - 5}
          width={w + 10}
          height={10}
          fill="none"
          pointerEvents="visible"
          onPointerDown={this._handleBrushStart}
          onPointerMove={event => {
            if (this.move) {
              const [x, y] = this.props.getEventMouse(event);
              const [sx, sy] = this.move;
              const dy = y - sy;
              const my = ybf(y1 + dy);
              const [my0, my1] = y0 < my ? [y0, my] : [y0, y0];
              let selection = this.state.selection;
              switch (brushType) {
                case '2d':
                case 'y':
                  selection = [
                    [x0, my0],
                    [x1, my1]
                  ];
              }
              this.move = [x, y];
              this.setState({selection});
              this.props.onBrush({
                target: this,
                type: 'brush',
                selection,
                sourceEvent: event
              });
              if (y0 >= my) {
                this.handleN.setPointerCapture(event.pointerId);
                return;
              }
            }
          }}
          onPointerUp={this._handleBrushEnd}
        />
        <rect
          ref={input => (this.handleW = input)}
          className="handle handle--w"
          cursor="ew-resize"
          x={x - 5}
          y={y - 5}
          width={10}
          height={h + 10}
          fill="none"
          pointerEvents="visible"
          onPointerDown={this._handleBrushStart}
          onPointerMove={event => {
            if (this.move) {
              const [x, y] = this.props.getEventMouse(event);
              const [sx, sy] = this.move;
              const dx = x - sx;
              const mx = xbf(x0 + dx);
              const [mx0, mx1] = mx < x1 ? [mx, x1] : [x1, x1];
              let selection = this.state.selection;
              switch (brushType) {
                case '2d':
                case 'x':
                  selection = [
                    [mx0, y0],
                    [mx1, y1]
                  ];
              }
              this.move = [x, y];
              this.setState({selection});
              this.props.onBrush({
                target: this,
                type: 'brush',
                selection,
                sourceEvent: event
              });
              if (mx >= x1) {
                this.handleE.setPointerCapture(event.pointerId);
                return;
              }
            }
          }}
          onPointerUp={this._handleBrushEnd}
        />
        <rect
          ref={input => (this.handleNW = input)}
          className="handle handle--nw"
          cursor="nwse-resize"
          x={x - 5}
          y={y - 5}
          width={10}
          height={10}
          fill="none"
          pointerEvents="visible"
          onPointerDown={this._handleBrushStart}
          onPointerMove={event => {
            if (this.move) {
              const [x, y] = this.props.getEventMouse(event);
              const [sx, sy] = this.move;
              const [dx, dy] = [x - sx, y - sy];
              let [mx, my] = [x0, y0];
              let mx0, mx1, my0, my1;
              let selection = this.state.selection;
              switch (brushType) {
                case '2d':
                  [mx, my] = [xbf(x0 + dx), ybf(y0 + dy)];
                  [mx0, mx1] = mx < x1 ? [mx, x1] : [x1, x1];
                  [my0, my1] = my < y1 ? [my, y1] : [y1, y1];
                  selection = [
                    [mx0, my0],
                    [mx1, my1]
                  ];
                  break;
                case 'x':
                  [mx, my] = [xbf(x0 + dx), y0];
                  [mx0, mx1] = mx < x1 ? [mx, x1] : [x1, x1];
                  selection = [
                    [mx0, y0],
                    [mx1, y1]
                  ];
                  break;
                case 'y':
                  [mx, my] = [x0, ybf(y0 + dy)];
                  [my0, my1] = my < y1 ? [my, y1] : [y1, y1];
                  selection = [
                    [x0, my0],
                    [x1, my1]
                  ];
              }
              this.move = [x, y];
              this.setState({selection});
              this.props.onBrush({
                target: this,
                type: 'brush',
                selection,
                sourceEvent: event
              });
              if (mx >= x1 && my >= y1) {
                this.handleSE.setPointerCapture(event.pointerId);
                return;
              } else if (mx >= x1) {
                this.handleNE.setPointerCapture(event.pointerId);
                return;
              } else if (my >= y1) {
                this.handleSW.setPointerCapture(event.pointerId);
                return;
              }
            }
          }}
          onPointerUp={this._handleBrushEnd}
        />
        <rect
          ref={input => (this.handleNE = input)}
          className="handle handle--ne"
          cursor="nesw-resize"
          x={x + w - 5}
          y={y - 5}
          width={10}
          height={10}
          fill="none"
          pointerEvents="visible"
          onPointerDown={this._handleBrushStart}
          onPointerMove={event => {
            if (this.move) {
              const [x, y] = this.props.getEventMouse(event);
              const [sx, sy] = this.move;
              const [dx, dy] = [x - sx, y - sy];
              let [mx, my] = [x1, y0];
              let mx0, mx1, my0, my1;
              let selection = this.state.selection;
              switch (brushType) {
                case '2d':
                  [mx, my] = [xbf(x1 + dx), ybf(y0 + dy)];
                  [mx0, mx1] = x0 < mx ? [x0, mx] : [x0, x0];
                  [my0, my1] = my < y1 ? [my, y1] : [y1, y1];
                  selection = [
                    [mx0, my0],
                    [mx1, my1]
                  ];
                  break;
                case 'x':
                  [mx, my] = [xbf(x1 + dx), y0];
                  [mx0, mx1] = x0 < mx ? [x0, mx] : [x0, x0];
                  selection = [
                    [mx0, y0],
                    [mx1, y1]
                  ];
                  break;
                case 'y':
                  [mx, my] = [x1, ybf(y0 + dy)];
                  [my0, my1] = my < y1 ? [my, y1] : [y1, y1];
                  selection = [
                    [x0, my0],
                    [x1, my1]
                  ];
              }
              this.move = [x, y];
              this.setState({selection});
              this.props.onBrush({
                target: this,
                type: 'brush',
                selection,
                sourceEvent: event
              });
              if (x0 >= mx && my >= y1) {
                this.handleSW.setPointerCapture(event.pointerId);
                return;
              } else if (x0 >= mx) {
                this.handleNW.setPointerCapture(event.pointerId);
                return;
              } else if (my >= y1) {
                this.handleSE.setPointerCapture(event.pointerId);
                return;
              }
            }
          }}
          onPointerUp={this._handleBrushEnd}
        />
        <rect
          ref={input => (this.handleSE = input)}
          className="handle handle--se"
          cursor="nwse-resize"
          x={x + w - 5}
          y={y + h - 5}
          width={10}
          height={10}
          fill="none"
          pointerEvents="visible"
          onPointerDown={this._handleBrushStart}
          onPointerMove={event => {
            if (this.move) {
              const [x, y] = this.props.getEventMouse(event);
              const [sx, sy] = this.move;
              const [dx, dy] = [x - sx, y - sy];
              let [mx, my] = [x1, y1];
              let mx0, mx1, my0, my1;
              let selection = this.state.selection;
              switch (brushType) {
                case '2d':
                  [mx, my] = [xbf(x1 + dx), ybf(y1 + dy)];
                  [mx0, mx1] = x0 < mx ? [x0, mx] : [x0, x0];
                  [my0, my1] = y0 < my ? [y0, my] : [y0, y0];
                  selection = [
                    [mx0, my0],
                    [mx1, my1]
                  ];
                  break;
                case 'x':
                  [mx, my] = [xbf(x1 + dx), y1];
                  [mx0, mx1] = x0 < mx ? [x0, mx] : [x0, x0];
                  selection = [
                    [mx0, y0],
                    [mx1, y1]
                  ];
                  break;
                case 'y':
                  [mx, my] = [x1, ybf(y1 + dy)];
                  [my0, my1] = y0 < my ? [y0, my] : [y0, y0];
                  selection = [
                    [x0, my0],
                    [x1, my1]
                  ];
              }
              this.move = [x, y];
              this.setState({selection});
              this.props.onBrush({
                target: this,
                type: 'brush',
                selection,
                sourceEvent: event
              });
              if (x0 >= mx && y0 >= my) {
                this.handleNW.setPointerCapture(event.pointerId);
                return;
              } else if (x0 >= mx) {
                this.handleSW.setPointerCapture(event.pointerId);
                return;
              } else if (y0 >= my) {
                this.handleNE.setPointerCapture(event.pointerId);
                return;
              }
            }
          }}
          onPointerUp={this._handleBrushEnd}
        />
        <rect
          ref={input => (this.handleSW = input)}
          className="handle handle--sw"
          cursor="nesw-resize"
          x={x - 5}
          y={y + h - 5}
          width={10}
          height={10}
          fill="none"
          pointerEvents="visible"
          onPointerDown={this._handleBrushStart}
          onPointerMove={event => {
            if (this.move) {
              const [x, y] = this.props.getEventMouse(event);
              const [sx, sy] = this.move;
              const [dx, dy] = [x - sx, y - sy];
              let [mx, my] = [x0, y1];
              let mx0, mx1, my0, my1;
              let selection = this.state.selection;
              switch (brushType) {
                case '2d':
                  [mx, my] = [xbf(x0 + dx), ybf(y1 + dy)];
                  [mx0, mx1] = mx < x1 ? [mx, x1] : [x1, x1];
                  [my0, my1] = y0 < my ? [y0, my] : [y0, y0];
                  selection = [
                    [mx0, my0],
                    [mx1, my1]
                  ];
                  break;
                case 'x':
                  [mx, my] = [xbf(x0 + dx), y1];
                  [mx0, mx1] = mx < x1 ? [mx, x1] : [x1, x1];
                  selection = [
                    [mx0, y0],
                    [mx1, y1]
                  ];
                  break;
                case 'y':
                  [mx, my] = [x0, ybf(y1 + dy)];
                  [my0, my1] = y0 < my ? [y0, my] : [y0, y0];
                  selection = [
                    [x0, my0],
                    [x1, my1]
                  ];
              }
              this.move = [x, y];
              this.setState({selection});
              this.props.onBrush({
                target: this,
                type: 'brush',
                selection,
                sourceEvent: event
              });
              if (mx >= x1 && y0 >= my) {
                this.handleNE.setPointerCapture(event.pointerId);
                return;
              } else if (mx >= x1) {
                this.handleSE.setPointerCapture(event.pointerId);
                return;
              } else if (y0 >= my) {
                this.handleNW.setPointerCapture(event.pointerId);
                return;
              }
            }
          }}
          onPointerUp={this._handleBrushEnd}
        />
      </React.Fragment>
    );
  }

  render() {
    return (
      <g className="brush">
        {this._renderOverlay()}
        {this._renderSelection()}
      </g>
    );
  }
}

export default SVGBrush;
export {SVGBrush};
